### DISCUSSION TITLE
NAT性能问题

### DISCUSSION THREAD

**Discussion by ysc3839** [immortalwrt/immortalwrt#1195] [Q&A] [answered]
最近我在调查路由器的 NAT 性能问题，路由器是 ipq40xx 的竞斗云2.0。
测试发现只有 ImmortalWrt 21.02.7 在开启 flow offloading 的情况下能跑满千兆，后续版本的 OpenWrt 和 ImmortalWrt (除了 snapshot) 不管是否开关 flow offloading 都只能跑到 200Mbps 左右，ImmortalWrt snapshot 能到 500Mbps，但离千兆仍然有距离。
请问有什么解决方案吗？是只能停留在旧版本？是否是因为新版本改用 nftables 导致的？后续版本有修复的可能性吗？

**Comment by 1715173329** [comment, marked as answer]
最新的 ImmortalWrt 23.05-SNAPSHOT 应该可以跑快一点，release 中的版本仍未修复 software flow offloading 问题。

21.02 及之前的交换机驱动应该是基于 QCA SDK 改的，自 23.05 起 ipq40xx 切换为 robimarko 编写的新 DSA 驱动，DSA 的效率目前来看本身就要比 Switch 低一些，并且这个新驱动为了符合规范，抛弃了一些 hack，性能进一步降低。
虽然 robimarko 本人说是以后有空会改善这一部分，但目前来看是遥遥无期。

**Comment by ysc3839** [comment]
@1715173329 感谢回复。
所以说是 DSA 驱动的锅更大？不是 nft flow offloading 的问题？
能否通过自行编译系统，把新版本的 DSA 换成旧版的交换机驱动呢？

**Reply by 1715173329** [reply]
nft flow offloading 上游 bug 导致软件加速这部分烂了，所以只能跑 200 Mbps。
除此之外是和 DSA 驱动有很大关系的。

> 能否通过自行编译系统，把新版本的 DSA 换成旧版的交换机驱动呢？

可以试试，但过程可能不是很愉快。

**Reply by ysc3839** [reply]
所以说 nft flow offloading 和 DSA 都有问题？那我也不太想尝试了，如果只是 DSA 的问题那试试倒无所谓。

**Reply by 1715173329** [reply]
nft 目前我们打上了 workaround，可以工作。

**Reply by ysc3839** [reply]
好的谢谢
